File: ContentProtectionSampleApp/ContentProtectionSampleApp/Utils/BasicEmbedTokenGenerator.m
//
//  BasicEmbedTokenGenerator.m
//  ContentProtectionSampleApp
//
//  Created on 9/1/16.
//  Copyright © 2016 Ooyala. All rights reserved.
//

#import "BasicEmbedTokenGenerator.h"
#import <OoyalaSDK/OOEmbeddedSecureURLGenerator.h>

@interface BasicEmbedTokenGenerator ()

- (instancetype)initWithPcode:(NSString *)pcode
                       apiKey:(NSString *)apiKey
                    apiSecret:(NSString *)apiSecret
                    accountId:(NSString *)accountId
                authorizeHost:(NSString *)authorizeHost;

@end

@implementation BasicEmbedTokenGenerator

- (instancetype)initWithPcode:(NSString *)pcode
                       apiKey:(NSString *)apiKey
                    apiSecret:(NSString *)apiSecret
                    accountId:(NSString *)accountId
                authorizeHost:(NSString *)authorizeHost {
  if (self = [super init]) {
    _pcode = pcode;
    _apiKey = apiKey;
    _apiSecret = apiSecret;
    _accountId = accountId;
    _authorizeHost = authorizeHost;
  }
  return self;
}

+ (instancetype)initWithPcode:(NSString *)pcode
                       apiKey:(NSString *)apiKey
                    apiSecret:(NSString *)apiSecret
                    accountId:(NSString *)accountId
                authorizeHost:(NSString *)authorizeHost {
  BasicEmbedTokenGenerator *tokenGenerator = [[BasicEmbedTokenGenerator alloc] initWithPcode:pcode
                                                                                      apiKey:apiKey
                                                                                   apiSecret:apiSecret
                                                                                   accountId:accountId
                                                                               authorizeHost:authorizeHost];
  return tokenGenerator;
}

- (void)tokenForEmbedCodes:(NSArray<NSString *> *)embedCodes
                  callback:(OOEmbedTokenCallback)callback {
  NSDictionary *params = @{@"account_id": self.accountId};
  NSString *uri = [NSString stringWithFormat:@"/sas/embed_token/%@/%@",
                   self.pcode, [embedCodes componentsJoinedByString:@","]];
  
  // You should not be using OOEmbeddedSecureURLGenerator in your own app.
  // We recommend generating this URL in a remote server you own, as we discourage storing apiKey and apiSecret information within the app because of security concerns.
  OOEmbeddedSecureURLGenerator* urlGen = [[OOEmbeddedSecureURLGenerator alloc] initWithAPIKey:self.apiKey
                                                                                       secret:self.apiSecret];
  NSURL* embedTokenUrl = [urlGen secureURLForHost:self.authorizeHost uri:uri params:params];
  callback(embedTokenUrl.absoluteString);
}

@end

File: PlaybackLab/MultiplePlayersScrollingViewSampleApp/MultiplePlayersScrollingViewSampleApp/Utils/BasicEmbedTokenGenerator.swift
//
//  BasicEmbedTokenGenerator.swift
//  MultiplePlayersScrollingViewSampleApp
//
//  Copyright © 2019 Ooyala. All rights reserved.
//

import UIKit

class BasicEmbedTokenGenerator: NSObject, OOEmbedTokenGenerator {

  public private(set) var pcode: String
  
  public private(set) var apiKey: String
  
  public private(set) var apiSecret: String
  
  public private(set) var accountId: String
  
  public private(set) var authorizeHost: String
  
  init(pcode: String, apiKey: String, apiSecret: String, accountId: String, authorizeHost: String) {
    self.pcode = pcode
    self.apiKey = apiKey
    self.apiSecret = apiSecret
    self.accountId = accountId
    self.authorizeHost = authorizeHost
    super.init()
  }
  
  func token(forEmbedCodes embedCodes: [Any]!, callback: OOEmbedTokenCallback!) {
    var params = [AnyHashable: Any]()
    params["account_id"] = accountId
    
    let uri = "/sas/embed_token/\(pcode)/\((embedCodes as! [String]).joined(separator: ","))"
    // You should not be using OOEmbeddedSecureURLGenerator in your own app.
    // We recommend generating this URL in a remote server you own, as we discourage storing apiKey and apiSecret information within the app because of security concerns.
    
    let urlGen = OOEmbeddedSecureURLGenerator(apiKey: apiKey, secret: apiSecret)!
    let embedTokenUrl = urlGen.secureURL(authorizeHost, uri: uri, params: params)!
    
    callback(embedTokenUrl.absoluteString)
  }
  
}
